src/cowboy_http2.erl

```diff
@@ -335,11 +353,42 @@ frame_rate(State0=#state{opts=Opts, frame_rate_num=Num0, frame_rate_time=Time},
 
 %% Frames received.
 
+%% Ignore frames for client-initiated streams as indicated in our GOAWAY.
```

Comment:
To be honest I'm starting to think that this should be handled directly in `cow_http2_machine`, not in Cowboy/Gun. So instead of only getting the LastStreamID when we initiate the close we would tell `cow_http2_machine` we are closing, and it will automatically ignore frames coming on later streams (perhaps with a limit on how much we are willing to ignore, later).

Reply:
OK, then `cow_http2_machine:frame/2` would just return `{ok, MachineState}` for ignored frames, I assume.

But `cow_http2_machine` would have to know we are closing, which means we need new field in `#http2_machine{}`?

Reply:
I was going to say add a value to `state`, but if we do that we won't be able to catch CONTINUATION errors. We still want to detect all protocol errors even if we are closing.

So yes, I suppose a new field is necessary. Maybe we can have a `remote_last_streamid` field that initially defaults to the max value (so no stream is ignored) and is set to the `remote_streamid` value when we send the second GOAWAY frame. So technically `cow_http2_machine` won't need to know about the details of the graceful shutdown implementation, it'll just know that it must not return frames above that value.

The function updating `remote_last_streamid` can return the value at the same time so a GOAWAY can be sent without having to call the `get_` function. We should also make sure that `cow_http2_machine` crashes if the user tries to change the value to a higher number since that's not allowed, at least as far as GOAWAYs are concerned. This'll prevent errors.

Reply:
Great!

If we always want to set `remote_last_streamid` to the current value of `remote_streamid`, there is no need to pass a value. Only passing the state would be enough. Then name `set_last_streamid/1` is weird if it doesn't take the value as a parameter, so maybe a better name would be `assign_last_streamid` or something like `stop_accepting_new_streams/1`?

This means `get_last_streamid/1` is obsoleted, I guess. I don't think it's used for anything else. Maybe it should be deprecated?

Another option is to add an `set_last_streamid/2` which just returns the new state (or `error(badarg)`). If you pass the value, it doesn't need to be returned, does it?

Reply:
I think `set_last_streamid/1` is fine because it doesn't preclude adding a second argument later on to set a custom value, for example if an implementation has good reasons to want to reduce the value even further. It also does what it says, only using a value found in the state.

I don't think `get_last_streamid/1` is obsolete, because we still need it for connection errors (Cowboy's `terminate` function for example). No need to update the HTTP2Machine state in that case. Anyway there's no need to deprecate the function because it's not documented to begin with.